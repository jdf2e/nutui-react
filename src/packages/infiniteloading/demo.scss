.demo-infinite {
  min-height: 100vh;
}
.infiniteUl {
  height: 300px;
  width: 100%;
  padding: 0;
  overflow-y: auto;
  overflow-x: hidden;
}
.infiniteLi {
  margin-top: 10px;
  font-size: 14px;
  color: rgba(100, 100, 100, 1);
  text-align: center;
}
.loading {
  display: block;
  width: 100%;
  text-align: center;
}
