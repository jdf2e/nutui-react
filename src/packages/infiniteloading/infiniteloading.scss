@import '../icon/icon.scss';
.nut-theme-dark {
  .nut-infiniteloading {
    .nut-infinite-top {
      .top-text {
        color: $dark-color;
      }
    }

    .nut-infinite-bottom {
      color: $dark-color;

      .bottom-text {
        color: $dark-color;
      }
    }
  }
}

.nut-infiniteloading {
  display: block;
  width: 100%;
  .nut-infinite-top {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    overflow: hidden;
    .top-box {
      width: 100%;
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;

      .top-img {
        width: 28px;
        height: 24px;
      }
      .top-text {
        font-size: 10px;
        color: $text-color;
      }
    }
  }
  .nut-infinite-bottom {
    display: block;
    width: 100%;
    padding-top: 16px;
    font-size: $font-size-small;
    color: $infinite-bottom-color;
    text-align: center;

    .bottom-box {
      .bottom-img {
        margin-right: 5px;
        width: 28px;
        height: 24px;
      }
      .bottom-text {
        font-size: 10px;
        color: $text-color;
      }
    }
  }
}
