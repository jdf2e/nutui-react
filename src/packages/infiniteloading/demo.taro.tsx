import React, { useState, useEffect, useMemo } from 'react'
import Taro from '@tarojs/taro'
import { Loading, More } from '@nutui/icons-react-taro'
import { useTranslate } from '@/sites/assets/locale/taro'
import { InfiniteLoading, Cell } from '@/packages/nutui.react.taro'
import '@/packages/infiniteloading/demo.scss'
import Header from '@/sites/components/header'
import { sleep } from '@/utils/sleep'
import { InfiniteLoadingStatusType } from './infiniteloading.taro'

interface T {
  '83913e71': string
  '84aa6bce': string
  eb4236fe: string
  '1254a90a': string
}

const InfiniteLoadingDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      '83913e71': '刷新成功',
      '84aa6bce': '基础演示',
      eb4236fe: '下拉刷新',
      '1254a90a': '没有啦~',
    },
    'zh-TW': {
      '83913e71': '刷新成功',
      '84aa6bce': '基礎演示',
      eb4236fe: '下拉刷新',
      '1254a90a': '沒有啦~',
    },
    'en-US': {
      '83913e71': 'Refresh successfully',
      '84aa6bce': 'Basic demo',
      eb4236fe: 'Pull down to refresh',
      '1254a90a': 'nope~',
    },
  })

  const [defaultList, setDefaultList] = useState<string[]>([])
  const [status, setStatus] = useState<InfiniteLoadingStatusType>('load')
  const [reverse] = useState(false)

  useEffect(() => {
    init()
  }, [])

  const loadMore = async () => {
    await sleep(2000)
    const curLen = defaultList.length
    const defaultCacheList = [...defaultList]
    if (defaultCacheList.length >= 100) {
      setStatus('loadMore')
    } else {
      for (let i = curLen; i < curLen + 10; i++) {
        defaultCacheList.push(`${i}`)
      }
      setDefaultList(defaultCacheList)
    }
  }

  const refresh = async () => {
    await sleep(2000)
    Taro.showToast({
      title: translated['83913e71'],
      icon: 'success',
      duration: 2000,
    })
  }

  const init = () => {
    for (let i = 0; i < 20; i++) {
      defaultList.push(`${i}`)
    }
    setDefaultList([...defaultList])
  }

  const fillColor: string = useMemo(() => {
    return reverse ? '#FFFFFF' : '#8C8C8C'
  }, [reverse])

  return (
    <>
      <Header />
      <div
        className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''} demo-infinite`}
      >
        <h2>{translated['84aa6bce']}</h2>
        <Cell>
          <ul
            className="infiniteUl"
            id="scrollDemo"
            style={{ height: '500px' }}
          >
            <InfiniteLoading
              pullRefresh
              target="scrollDemo"
              status={status}
              onLoadMore={loadMore}
              reverse={reverse}
              onRefresh={refresh}
              pullingText={
                <>
                  <Loading
                    className="nut-infiniteloading-top-tips-icons"
                    color={fillColor}
                  />
                  松开刷新
                </>
              }
              loadingText={
                <>
                  <Loading
                    className="nut-infiniteloading-bottom-tips-icons"
                    color={fillColor}
                  />
                  加载中
                </>
              }
              loadMoreText={
                <>
                  <More
                    className="nut-infiniteloading-bottom-tips-icons"
                    color={fillColor}
                  />
                  没有更多了
                </>
              }
            >
              {defaultList.map((item, index) => {
                return (
                  <li className="infiniteLi" key={index}>
                    {item}
                  </li>
                )
              })}
            </InfiniteLoading>
          </ul>
        </Cell>
      </div>
    </>
  )
}

export default InfiniteLoadingDemo
