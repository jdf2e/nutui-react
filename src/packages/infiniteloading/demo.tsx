import React, { useState, useEffect } from 'react'
import { Jd } from '@nutui/icons-react'
import { useTranslate } from '../../sites/assets/locale'
import { InfiniteLoading } from './infiniteloading'
import Cell from '@/packages/cell'
import Toast from '@/packages/toast'
import './demo.scss'

interface T {
  '83913e71': string
  '84aa6bce': string
  eb4236fe: string
  '9ed40460': string
  '1254a90a': string
  '1254a90n': string
}
const InfiniteloadingDemo = () => {
  const [translated] = useTranslate<T>({
    'zh-CN': {
      '83913e71': '刷新成功',
      '84aa6bce': '基础用法',
      eb4236fe: '下拉刷新',
      '9ed40460': '自定义加载文案',
      '1254a90a': '没有啦~',
      '1254a90n': '基于window滚动',
    },
    'zh-TW': {
      '83913e71': '刷新成功',
      '84aa6bce': '基礎用法',
      eb4236fe: '下拉刷新',
      '9ed40460': '自定義加載文案',
      '1254a90a': '沒有啦~',
      '1254a90n': '基於window滾動',
    },
    'en-US': {
      '83913e71': 'Refresh successfully',
      '84aa6bce': 'Basic usage',
      eb4236fe: 'Pull down to refresh',
      '9ed40460': 'custom loading text',
      '1254a90a': 'nope~',
      '1254a90n': 'window scroll',
    },
  })

  const [defaultList, setDefaultList] = useState<string[]>([])
  const [customList, setCustomList] = useState<string[]>([])
  const [refreshList, setRefreshList] = useState<string[]>([])
  const [windowList, setWindowList] = useState<string[]>([])
  const [hasMore, setHasMore] = useState(true)
  const [customHasMore, setCustomHasMore] = useState(true)
  const [refreshHasMore, setRefreshHasMore] = useState(true)
  const [windownHasMore, setWindowHasMore] = useState(true)

  useEffect(() => {
    init()
  }, [])

  const loadMore = (done: () => void) => {
    setTimeout(() => {
      const curLen = defaultList.length
      for (let i = curLen; i < curLen + 10; i++) {
        defaultList.push(`${i}`)
      }
      if (defaultList.length >= 30) {
        setHasMore(false)
      } else {
        setDefaultList([...defaultList])
      }
      done()
    }, 500)
  }

  const refreshLoadMore = (done: () => void) => {
    setTimeout(() => {
      const curLen = refreshList.length
      for (let i = curLen; i < curLen + 10; i++) {
        refreshList.push(`${i}`)
      }
      if (refreshList.length >= 30) {
        setRefreshHasMore(false)
      } else {
        setRefreshList([...refreshList])
      }
      done()
    }, 500)
  }

  const customLoadMore = (done: () => void) => {
    setTimeout(() => {
      const curLen = customList.length
      for (let i = curLen; i < curLen + 10; i++) {
        customList.push(`${i}`)
      }
      if (customList.length >= 30) {
        setCustomHasMore(false)
      } else {
        setCustomList([...customList])
      }
      done()
    }, 500)
  }

  const windowLoadMore = (done: () => void) => {
    setTimeout(() => {
      const curLen = windowList.length
      for (let i = curLen; i < curLen + 10; i++) {
        windowList.push(`${i}`)
      }
      if (windowList.length >= 300) {
        setWindowHasMore(false)
      } else {
        setWindowList([...windowList])
      }
      done()
    }, 500)
  }

  const refresh = (done: () => void) => {
    setTimeout(() => {
      Toast.show(translated['83913e71'])
      done()
    }, 1000)
  }

  const init = () => {
    for (let i = 0; i < 10; i++) {
      defaultList.push(`${i}`)
      customList.push(`${i}`)
      refreshList.push(`${i}`)
      windowList.push(`${i}`)
    }
    setDefaultList([...defaultList])
    setCustomList([...customList])
    setRefreshList([...refreshList])
    setRefreshList([...windowList])
  }

  return (
    <>
      <div className="demo">
        <h2>{translated['84aa6bce']}</h2>
        <Cell>
          <ul className="infiniteUl" id="scroll">
            <InfiniteLoading
              target="scroll"
              hasMore={hasMore}
              onLoadMore={loadMore}
            >
              {defaultList.map((item, index) => {
                return (
                  <li className="infiniteLi" key={index}>
                    {item}
                  </li>
                )
              })}
            </InfiniteLoading>
          </ul>
        </Cell>

        <h2>{translated.eb4236fe}</h2>
        <Cell>
          <ul className="infiniteUl" id="refreshScroll">
            <InfiniteLoading
              pullingText={
                <>
                  <Jd />
                  <span style={{ fontSize: '10px' }}>松开刷新</span>
                </>
              }
              loadingText={<Jd />}
              target="refreshScroll"
              pullRefresh
              hasMore={refreshHasMore}
              onLoadMore={refreshLoadMore}
              onRefresh={refresh}
            >
              {refreshList.map((item, index) => {
                return (
                  <li className="infiniteLi" key={index}>
                    {item}
                  </li>
                )
              })}
            </InfiniteLoading>
          </ul>
        </Cell>

        <h2>{translated['9ed40460']}</h2>
        <Cell>
          <ul className="infiniteUl" id="customScroll">
            <InfiniteLoading
              target="customScroll"
              loadingText="loading"
              loadMoreText={translated['1254a90a']}
              hasMore={customHasMore}
              onLoadMore={customLoadMore}
            >
              {customList.map((item, index) => {
                return (
                  <li className="infiniteLi" key={index}>
                    {item}
                  </li>
                )
              })}
            </InfiniteLoading>
          </ul>
        </Cell>

        <h2>{translated['1254a90n']}</h2>
        <Cell>
          <ul className="infiniteUl2">
            <InfiniteLoading
              loadingText="loading"
              loadMoreText={translated['1254a90a']}
              hasMore={windownHasMore}
              onLoadMore={windowLoadMore}
            >
              {windowList.map((item, index) => {
                return (
                  <li className="infiniteLi" key={index}>
                    {item}
                  </li>
                )
              })}
            </InfiniteLoading>
          </ul>
        </Cell>
      </div>
    </>
  )
}

export default InfiniteloadingDemo
