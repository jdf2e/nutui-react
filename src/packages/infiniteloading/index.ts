import { Infiniteloading } from './infiniteloading'

export default Infiniteloading
