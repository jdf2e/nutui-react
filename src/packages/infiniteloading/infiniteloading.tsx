import React, {
  useState,
  useEffect,
  useRef,
  FunctionComponent,
  ReactNode,
} from 'react'
import classNames from 'classnames'
import { useConfig } from '@/packages/configprovider'
import { BasicComponent, ComponentDefaults } from '@/utils/typings'
import { InfiniteLoadingType } from './types'

export interface InfiniteLoadingProps extends BasicComponent {
  type: InfiniteLoadingType
  hasMore: boolean
  threshold: number
  target: string
  capture: boolean
  pullRefresh: boolean
  pullingText: ReactNode
  loadingText: ReactNode
  loadMoreText: ReactNode
  onRefresh: () => Promise<void>
  onLoadMore: () => Promise<void>
  onScroll: (param: number) => void
}

declare let window: Window & { webkitRequestAnimationFrame: any } & {
  mozRequestAnimationFrame: any
}

const defaultProps = {
  ...ComponentDefaults,
  type: 'default',
  hasMore: true,
  threshold: 200,
  target: '',
  capture: false,
  pullRefresh: false,
} as InfiniteLoadingProps

const classPrefix = `nut-infiniteloading`
export const InfiniteLoading: FunctionComponent<
  Partial<InfiniteLoadingProps> &
    Omit<React.HTMLAttributes<HTMLDivElement>, 'onScroll'>
> = (props) => {
  const { locale } = useConfig()
  const {
    children,
    type,
    hasMore,
    threshold,
    target,
    capture,
    pullRefresh,
    pullingText,
    loadingText,
    loadMoreText,
    className,
    onRefresh,
    onLoadMore,
    onScroll,
    ...restProps
  } = {
    ...defaultProps,
    ...props,
  }
  const [isInfiniting, setIsInfiniting] = useState(false)
  const scroller = useRef<HTMLDivElement>(null)
  const refreshTop = useRef<HTMLDivElement>(null)
  const scrollEl = useRef<Window | HTMLElement | null>(null)
  const isTouching = useRef(false)
  const beforeScrollTop = useRef(0)
  const refreshMaxH = useRef(0)
  const y = useRef(0)
  const distance = useRef(0)

  const classes = classNames(classPrefix, className, `${classPrefix}-${type}`)

  useEffect(() => {
    if (target && document.getElementById(target)) {
      scrollEl.current = document.getElementById(target)
    } else {
      scrollEl.current = window
    }
    scrollEl.current?.addEventListener('scroll', handleScroll, capture)

    return () => {
      scrollEl.current?.removeEventListener('scroll', handleScroll, capture)
    }
  }, [hasMore, isInfiniting, onLoadMore])

  useEffect(() => {
    const element = scroller.current as HTMLDivElement
    element.addEventListener('touchmove', touchMove, { passive: false })

    return () => {
      element.removeEventListener('touchmove', touchMove, {
        passive: false,
      } as EventListenerOptions)
    }
  }, [])

  const getStyle = () => {
    return {
      height: distance.current < 0 ? `0px` : `${distance.current}px`,
      transition: isTouching.current
        ? `height 0s cubic-bezier(0.25,0.1,0.25,1)`
        : `height 0.2s cubic-bezier(0.25,0.1,0.25,1)`,
    }
  }

  const handleScroll = async () => {
    if (!isScrollAtBottom() || !hasMore || isInfiniting) {
      return
    }
    setIsInfiniting(true)
    await onLoadMore?.()
    infiniteDone()
  }

  const infiniteDone = () => {
    setIsInfiniting(false)
  }

  const getRefreshTop = () => {
    return refreshTop.current as HTMLDivElement
  }

  const refreshDone = () => {
    distance.current = 0
    getRefreshTop().style.height = `${distance.current}px`
    isTouching.current = false
  }

  const touchStart = (event: React.TouchEvent<HTMLDivElement>) => {
    if (beforeScrollTop.current === 0 && !isTouching.current && pullRefresh) {
      y.current = event.touches[0].pageY
      isTouching.current = true
      const childHeight = (getRefreshTop().firstElementChild as HTMLElement)
        .offsetHeight
      refreshMaxH.current = Math.floor(childHeight * 1 + 10)
    }
  }

  const touchMove = (event: any) => {
    distance.current = event.touches[0].pageY - y.current
    if (distance.current > 0 && isTouching.current) {
      event.preventDefault()
      if (distance.current >= refreshMaxH.current) {
        distance.current = refreshMaxH.current
        getRefreshTop().style.height = `${distance.current}px`
      } else {
        getRefreshTop().style.height = `${distance.current}px`
      }
    } else {
      distance.current = 0
      getRefreshTop().style.height = `${distance.current}px`
      isTouching.current = false
    }
  }

  const touchEnd = async () => {
    if (distance.current < refreshMaxH.current) {
      distance.current = 0
      getRefreshTop().style.height = `${distance.current}px`
      isTouching.current = false
    } else {
      await onRefresh?.()
      refreshDone()
    }
  }

  const getWindowScrollTop = () => {
    return window.scrollY !== undefined
      ? window.scrollY
      : (document.documentElement || document.body.parentNode || document.body)
          .scrollTop
  }

  const calculateTopPosition = (el: HTMLElement): number => {
    return !el
      ? 0
      : el.offsetTop + calculateTopPosition(el.offsetParent as HTMLElement)
  }

  const isScrollAtBottom = () => {
    let offsetDistance = 0
    let resScrollTop = 0
    let direction = 'down'
    const windowScrollTop = getWindowScrollTop()
    if (!target || !document.getElementById(target)) {
      if (scroller.current) {
        offsetDistance =
          calculateTopPosition(scroller.current) +
          scroller.current.offsetHeight -
          windowScrollTop -
          window.innerHeight
      }
      resScrollTop = windowScrollTop
    } else {
      const { scrollHeight, clientHeight, scrollTop } =
        scrollEl.current as HTMLElement
      offsetDistance = scrollHeight - clientHeight - scrollTop
      resScrollTop = scrollTop
    }
    if (beforeScrollTop.current > resScrollTop) {
      direction = 'up'
    } else {
      direction = 'down'
    }
    beforeScrollTop.current = resScrollTop
    onScroll && onScroll(resScrollTop)
    return offsetDistance <= threshold && direction === 'down'
  }

  return (
    <div
      className={classes}
      ref={scroller}
      onTouchStart={touchStart}
      onTouchEnd={touchEnd}
      {...restProps}
    >
      <div className="nut-infinite-top" ref={refreshTop} style={getStyle()}>
        <div className="nut-infinite-top-tips">
          {pullingText || locale.infiniteloading.pullRefreshText}
        </div>
      </div>
      <div className="nut-infinite-container">{children}</div>
      <div className="nut-infinite-bottom">
        {isInfiniting ? (
          <div className="nut-infinite-bottom-tips">
            {loadingText || locale.infiniteloading.loadText}
          </div>
        ) : (
          !hasMore && (
            <div className="nut-infinite-bottom-tips">
              {loadMoreText || locale.infiniteloading.loadMoreText}
            </div>
          )
        )}
      </div>
    </div>
  )
}

InfiniteLoading.defaultProps = defaultProps
InfiniteLoading.displayName = 'NutInfiniteLoading'
