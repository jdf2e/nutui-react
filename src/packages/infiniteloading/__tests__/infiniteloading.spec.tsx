import * as React from 'react'
import { act, render, waitFor } from '@testing-library/react'
import { trigger, triggerDrag } from '@/utils/test/event'
import '@testing-library/jest-dom'

import { InfiniteLoading } from '../infiniteloading'
import { sleep } from '@/utils/sleep'

test('pull base', () => {
  const refresh = vi.fn()
  const { container } = render(
    <InfiniteLoading pullRefresh loadMoreText="没有更多" onRefresh={refresh} />
  )
  const track = container.querySelector('.nut-infiniteloading')

  // pulling
  trigger(track, 'touchstart', 0, 0)
  trigger(track, 'touchmove', 0, 20)
  expect(container).toMatchSnapshot()

  // loading
  trigger(track, 'touchend', 0, 100)
  expect(container).toMatchSnapshot()

  // still loading
  triggerDrag(track, 0, 100)
  expect(refresh).toBeCalled()
})

test('pull base 03', () => {
  const refresh = vi.fn()
  const { container } = render(
    <InfiniteLoading pullRefresh pullingText="下拉刷新" onRefresh={refresh} />
  )
  const track = container.querySelector('.nut-infiniteloading')

  // pulling
  trigger(track, 'touchstart', 0, 0)
  trigger(track, 'touchmove', 0, 0)
  expect(container).toMatchSnapshot()

  // loading
  trigger(track, 'touchend', 0, 5)
  expect(container).toMatchSnapshot()

  // still loading
  triggerDrag(track, 0, 5)
  expect(refresh).toHaveBeenCalledTimes(0)
})

test('infiniteloading base', () => {
  const { container } = render(
    <ul className="infiniteUl" id="scroll">
      <InfiniteLoading target="scroll" />
    </ul>
  )
})

test('infiniteloading base 01', async () => {
  const App = () => {
    const [refreshList, setRefreshList] = React.useState<string[]>([])
    const [refreshHasMore, setRefreshHasMore] = React.useState(true)

    React.useEffect(() => {
      init()
    }, [])

    const init = () => {
      for (let i = 0; i < 10; i++) {
        refreshList.push(`${i}`)
      }
      setRefreshList([...refreshList])
    }

    const refreshLoadMore = async () => {
      await sleep(10)
      const curLen = refreshList.length
      for (let i = curLen; i < curLen + 10; i++) {
        refreshList.push(`${i}`)
      }
      if (refreshList.length >= 300) {
        setRefreshHasMore(false)
      } else {
        setRefreshList([...refreshList])
      }
    }
    return (
      <InfiniteLoading
        loadMoreText="没有更多"
        onLoadMore={refreshLoadMore}
        hasMore={refreshHasMore}
      />
    )
  }
  const { container } = render(<App />)
  await waitFor(() => expect(container).toMatchSnapshot())
})

test('infiniteloading base 02', async () => {
  const done = vi.fn()
  const App = () => {
    const [refreshList, setRefreshList] = React.useState<string[]>([])
    const [refreshHasMore, setRefreshHasMore] = React.useState(true)

    React.useEffect(() => {
      init()
    }, [])

    const init = () => {
      for (let i = 0; i < 10; i++) {
        refreshList.push(`${i}`)
      }
      setRefreshList([...refreshList])
    }

    const refreshLoadMore = async () => {
      await sleep(100)
      const curLen = refreshList.length
      for (let i = curLen; i < curLen + 10; i++) {
        refreshList.push(`${i}`)
      }
      if (refreshList.length >= 30) {
        setRefreshHasMore(false)
      } else {
        setRefreshList([...refreshList])
      }
    }
    return (
      <InfiniteLoading
        loadMoreText="没有更多"
        onLoadMore={refreshLoadMore}
        hasMore={refreshHasMore}
        onScroll={done}
      >
        {refreshList.map((item, index) => {
          return (
            <li className="infiniteLi" key={index}>
              {item}
            </li>
          )
        })}
      </InfiniteLoading>
    )
  }
  const { container } = render(<App />)
  const track = container.querySelector('.nut-infiniteloading')
  trigger(track, 'scroll', 0, 800)
  expect(container).toMatchSnapshot()
  await waitFor(() => expect(done).toHaveBeenCalled())
})

test('hasMore false', () => {
  const done = vi.fn()
  const { container: container1, rerender } = render(
    <InfiniteLoading loadMoreText="没有更多" hasMore={false} onScroll={done}>
      {Array.from<string>({ length: 100 })
        .fill('NutUI')
        .map((item: string, index) => {
          return (
            <li className="infiniteLi" key={index}>
              {item}
            </li>
          )
        })}
    </InfiniteLoading>
  )
  const track1 = container1.querySelector('.nut-infiniteloading')
  trigger(track1, 'scroll', 0, 100)
})

test('hasMore', () => {
  const done = vi.fn()
  const { container } = render(
    <InfiniteLoading loadMoreText="没有更多" hasMore onScroll={done}>
      {Array.from<string>({ length: 100 })
        .fill('NutUI')
        .map((item: string, index) => {
          return (
            <li className="infiniteLi" key={index}>
              {item}
            </li>
          )
        })}
    </InfiniteLoading>
  )
  const track1 = container.querySelector('.nut-infiniteloading')
  act(() => {
    trigger(track1, 'scroll', 0, 100)
  })

  waitFor(() => {
    expect(done).toBeCalled()
  })
})

test('pull base 01', async () => {
  const refresh = async () => {
    await sleep(10)
  }
  const { container } = render(
    <InfiniteLoading pullRefresh pullingText="下拉刷新" onRefresh={refresh} />
  )
  const track = container.querySelector('.nut-infiniteloading')

  // pulling
  trigger(track, 'touchstart', 0, 0)
  trigger(track, 'touchmove', 0, 20)
  expect(container).toMatchSnapshot()

  // loading
  trigger(track, 'touchend', 0, 100)
  expect(container).toMatchSnapshot()
})
