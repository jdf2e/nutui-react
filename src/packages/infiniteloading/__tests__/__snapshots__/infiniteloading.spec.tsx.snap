// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`infiniteloading base 01 1`] = `
<div>
  <div
    class="nut-infiniteloading"
  >
    <div
      class="nut-infinite-top"
      style="height: 0px; transition: height 0.2s cubic-bezier(0.25,0.1,0.25,1);"
    >
      <div
        class="nut-infinite-top-tips"
      >
        松开刷新
      </div>
    </div>
    <div
      class="nut-infinite-container"
    />
    <div
      class="nut-infinite-bottom"
    />
  </div>
</div>
`;

exports[`infiniteloading base 02 1`] = `
<div>
  <div
    class="nut-infiniteloading"
  >
    <div
      class="nut-infinite-top"
      style="height: 0px; transition: height 0.2s cubic-bezier(0.25,0.1,0.25,1);"
    >
      <div
        class="nut-infinite-top-tips"
      >
        松开刷新
      </div>
    </div>
    <div
      class="nut-infinite-container"
    >
      <li
        class="infiniteLi"
      >
        0
      </li>
      <li
        class="infiniteLi"
      >
        1
      </li>
      <li
        class="infiniteLi"
      >
        2
      </li>
      <li
        class="infiniteLi"
      >
        3
      </li>
      <li
        class="infiniteLi"
      >
        4
      </li>
      <li
        class="infiniteLi"
      >
        5
      </li>
      <li
        class="infiniteLi"
      >
        6
      </li>
      <li
        class="infiniteLi"
      >
        7
      </li>
      <li
        class="infiniteLi"
      >
        8
      </li>
      <li
        class="infiniteLi"
      >
        9
      </li>
    </div>
    <div
      class="nut-infinite-bottom"
    />
  </div>
</div>
`;

exports[`pull base 01 1`] = `
<div>
  <div
    class="nut-infiniteloading"
  >
    <div
      class="nut-infinite-top"
      style="height: 10px; transition: height 0.2s cubic-bezier(0.25,0.1,0.25,1);"
    >
      <div
        class="nut-infinite-top-tips"
      >
        下拉刷新
      </div>
    </div>
    <div
      class="nut-infinite-container"
    />
    <div
      class="nut-infinite-bottom"
    />
  </div>
</div>
`;

exports[`pull base 01 2`] = `
<div>
  <div
    class="nut-infiniteloading"
  >
    <div
      class="nut-infinite-top"
      style="height: 10px; transition: height 0.2s cubic-bezier(0.25,0.1,0.25,1);"
    >
      <div
        class="nut-infinite-top-tips"
      >
        下拉刷新
      </div>
    </div>
    <div
      class="nut-infinite-container"
    />
    <div
      class="nut-infinite-bottom"
    />
  </div>
</div>
`;

exports[`pull base 03 1`] = `
<div>
  <div
    class="nut-infiniteloading"
  >
    <div
      class="nut-infinite-top"
      style="height: 0px; transition: height 0.2s cubic-bezier(0.25,0.1,0.25,1);"
    >
      <div
        class="nut-infinite-top-tips"
      >
        下拉刷新
      </div>
    </div>
    <div
      class="nut-infinite-container"
    />
    <div
      class="nut-infinite-bottom"
    />
  </div>
</div>
`;

exports[`pull base 03 2`] = `
<div>
  <div
    class="nut-infiniteloading"
  >
    <div
      class="nut-infinite-top"
      style="height: 0px; transition: height 0.2s cubic-bezier(0.25,0.1,0.25,1);"
    >
      <div
        class="nut-infinite-top-tips"
      >
        下拉刷新
      </div>
    </div>
    <div
      class="nut-infinite-container"
    />
    <div
      class="nut-infinite-bottom"
    />
  </div>
</div>
`;

exports[`pull base 1`] = `
<div>
  <div
    class="nut-infiniteloading"
  >
    <div
      class="nut-infinite-top"
      style="height: 10px; transition: height 0.2s cubic-bezier(0.25,0.1,0.25,1);"
    >
      <div
        class="nut-infinite-top-tips"
      >
        松开刷新
      </div>
    </div>
    <div
      class="nut-infinite-container"
    />
    <div
      class="nut-infinite-bottom"
    />
  </div>
</div>
`;

exports[`pull base 2`] = `
<div>
  <div
    class="nut-infiniteloading"
  >
    <div
      class="nut-infinite-top"
      style="height: 10px; transition: height 0.2s cubic-bezier(0.25,0.1,0.25,1);"
    >
      <div
        class="nut-infinite-top-tips"
      >
        松开刷新
      </div>
    </div>
    <div
      class="nut-infinite-container"
    />
    <div
      class="nut-infinite-bottom"
    />
  </div>
</div>
`;
